enhancements/ingress/bind-options.md

```diff
@@ -0,0 +1,295 @@
+        // +kubebuilder:default:=80
+        // +optional
+        HTTPPort int32 `json:"httpPort,omitempty"`
+
+        // httpsPort defines the port number which HAProxy process binds for
+        // https connections. Setting this field is generally not recommended. However in
+        // HostNetwork strategy, default https 443 port might be occupied by other processess
+        //
+        // +kubebuilder:validation:Optional
+        // +kubebuilder:validation:Minimum=1
+        // +kubebuilder:validation:Maximum=30000
+        // +kubebuilder:default:=443
+        // +optional
+        HTTPSPort int32 `json:"httpsPort,omitempty"`
+
+        // sniPort is for some internal front-end to back-end communication
+        // This port can be anything you want as long as they are unique on the machine.
+        // This port will not be exposed externally.
+        // +kubebuilder:validation:Optional
+        // +kubebuilder:validation:Minimum=1
+        // +kubebuilder:validation:Maximum=30000
+        // +kubebuilder:default:=10444
+        // +optional
+        SNIPort int32 `json:"sniPort,omitempty"`
+
+        // noSniPort is for some internal front-end to back-end communication
+        // This port can be anything you want as long as they are unique on the machine.
+        // This port will not be exposed externally.
+        // +kubebuilder:validation:Optional
+        // +kubebuilder:validation:Minimum=1
+        // +kubebuilder:validation:Maximum=30000
+        // +kubebuilder:default:=10443
+        // +optional
+        NoSNIPort int32 `json:"noSniPort,omitempty"`
+
+        // statsPort is the port number which HAProxy process binds
+        // to expose statistics on it. Setting this field is generally not recommended.
+        // However in HostNetwork strategy, default stats port 1936 might
+        // be occupied by other processess
+        //
+        // +kubebuilder:validation:Optional
+        // +kubebuilder:validation:Minimum=1
+        // +kubebuilder:validation:Maximum=30000
+        // +kubebuilder:default:=1936
+        // +optional
+        StatsPort int32 `json:"statsPort,omitempty"`
+    }
+    ```
+
+The following example configures two IngressControllers with HostNetwork strategy that can run on the same nodes without port conflicts:
+
+    ```yaml
+    ---
+    apiVersion: operator.openshift.io/v1
+    kind: IngressController
+    metadata:
+    name: default-1
+    namespace: openshift-ingress-operator
+    spec:
+    endpointPublishingStrategy:
+        type: HostNetwork
+    ---
+    apiVersion: operator.openshift.io/v1
+    kind: IngressController
+    metadata:
+    name: default-2
+    namespace: openshift-ingress-operator
+    spec:
+    endpointPublishingStrategy:
+        type: HostNetwork
+        hostNetwork:
+        bindOptions:
+            httpPort: 11080
+            httpsPort: 11443
+            sniPort: 11444
+            noSniPort: 11445
+            statsPort: 1937
```

Comment:
My only uncertainty about this API is whether we should make it more generic or not. I don't quite appreciate the repetition of `Port` in all the fields. I was thinking about 2 alternatives:
1. generic
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
name: default-2
namespace: openshift-ingress-operator
spec:
endpointPublishingStrategy:
    type: HostNetwork
    hostNetwork:
      bindOptions:
          ports:
            http: 11080
            https: 11443
            sni: 11444
            nosni: 11445
            stats: 1937
```
2. even more generic
```yaml
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
name: default-2
namespace: openshift-ingress-operator
spec:
endpointPublishingStrategy:
    type: HostNetwork
    hostNetwork:
      bindOptions:
          ports:
            - name: "http"
              port: 11080
            - name: "https"
              port: 11443
            - name: "sni"
              port: 11444
            - name: "nosni"
              port: 11445
            - name: "stats"
              port: 1937
```

What do you think @m-yosefpor ? You can push back if you disagree, it's perfectly fine. As a matter of fact I don't have a strong opinion on this, just wanted to discuss the alternatives.

cc @Miciah, @frobware I would like to know your opinion on the whole API and on this comment in particular.

Reply:
@alebedev87 I completely agree 👌 the "port" word repetition in the fields is undesirable. Option 1 definitely seems better than what currently is. For option 2,  it depends on the controller : for HAProxy, I think it is overkill as there are not any other ports other than those currently defined in router config, and I think it's very unlikely that they change in the near future. However, if the underlying controller changes to sth else, the more generic method in option 2 would be better. It also mitigates the part mentioned in the "risks" section about this.

Also a number 3 option would be sth like this:

```
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
name: default-2
namespace: openshift-ingress-operator
spec:
endpointPublishingStrategy:
    type: HostNetwork
    hostNetwork:
      portBindings:
        http: 11080
        https: 11443
        sni: 11444
        nosni: 11445
        stats: 1937
```
 
Is it likely to put anything else under `bindOptions` other than port numbers anywhere in the future? If you think there would not be any other parameter related to bindings other than port numbers in HostNetwork strategy, another option would be renaming the `bindOptions` instead. 

Or an option 4, the corresponding generic definition like option 2, with `bindOptions` renamed:

```
endpointPublishingStrategy:
    type: HostNetwork
    hostNetwork:
      portBindings:
        - name: "http"
          port: 11080
 ```

P.S. on controller implementation: I don't think the openshift community is going to change the ingressController implementation from HAProxy to contour (or anything else) ever, although they might use contour for supporting Gateway-api (formerly Service-api) and probably introducing new CRDs/COs for lifecycle of the controller. Am I right?

Reply:
Option 3 looks good to me too. I don't think we plan to do anything other than ports for `bindOptions`. However keeping the name more open for further enhancements sounds tempting. When I was looking at the API  I had `bind` Linux syscall in mind, from that point of view `bindOptions` looks quite self-contained and leave enough room for other fields (i.e. address).
About other implementations of the ingress controller, API was designed to be as generic as possible, even Contour can take the place of HAProxy, no concrete plans for the moment though.

Well, at the end I have to choose something, so I vote for the option 1. Also I would like to know what you guys (@Miciah, @frobware) think about the options.

Reply:
At the moment I would lean more to option 2. A list without hard-coded names which only make sense (ATM) for haproxy.

Reply:
@m-yosefpor, please find the outcome of the discussion @frobware @Miciah and I had on this API enhancement:

We have some concerns about putting the sni/nosni port configuration in the API:

- The sysadmin’s goal is really only to configure alternative ports for 80/443/1936 and to be able to point an external LB to these ports.
- Requiring the sysadmin to configure these additional ports would mean we would need additional documentation to explain why the sysadmin needed to care about something that is really an implementation detail of the HAProxy configuration.  (And this internal implementation detail could conceivably change, in which case the API would become cruft.)
- The larger API surface comes with an increased risk of misconfiguration and port conflicts that an automated solution could prevent.
- Requiring 5 ports instead of just 3 reduces scalability for users who want to run a large number of host-network router pods on the same node.
- Requiring HAProxy-specific configuration makes migration to another proxy (such as Contour) less straightforward.  

As an alternative to requiring the sysadmin to configure the sni/nosni ports, how about the following?

- The sysadmin configures alternatives for ports 80/443/1936.
- The cluster ingress operator configures the router to use these alternative ports, and configures HAProxy to use a unique loopback address for the sni/nosni frontends.  
     - This loopback address would be determined deterministically, for example by hashing the ingresscontroller’s name.  For example, the “default” ingresscontroller could use 127.1.2.3, and the “custom” ingresscontroller could use 127.4.5.6.
     - The sni/nosni port numbers wouldn’t need to change to avoid the conflict as the routers would be using different IP addresses.  

On further thought, the need to configure the binding address:port for the sni/nosni frontends raises the question of whether we need this internal hop over a TCP connection.  We’ll look into why we have that hop, and whether we can get rid of the loopback hop using a Unix domain socket or other solution, which would have other advantages as well.

Reply:
@alebedev87 Thank you very much for your detailed response. Great conclusions! Actually I thought about CIO to decide the ports for sni/nosni, but then I realized it would be very situation dependent and raises much more issues. The idea of listening on different lo addresses (and better, a unix domain) is awesome.

So I'll modify the proposal to include only the 3 mentioned ports. And for the format, @frobware was leaning towards option2, but now that we are going to remove sni/nosni, should we go with option 1 instead?

Reply:
Sorry @m-yosefpor, I was off for a week.
Yes, I think the option 1 with only 3 ports (http,https,stats) would be good.

Reply:
@alebedev87 sorry I was also off for a week (dealing with COVID). I'm going to start working on this again. Another challenge that we faced `X-Forward-Port` is set to the port the router is listening on. So if you change the port to `10443`, then `X-Forwarded-Port: 10443`. Most users have an external LB on standard port 80 and 443, and want to loadbalance to other ports (e.g. 10080, 10443). However with this header set, `image-registry` will stop working as it redirects users to `X-Forwarded-Port`, and the user tries to connect to this port of external LB which fails. There might be some other apps which will face this issue too. Should I also mention this in the enhancement? any changes needed in image-registry? or introducing a new option in routers to override X-Forwarded-Port to any custom value? (currently it is possible to set policy for forward headers in routes, but  it does not help with this issue, as the only option is to set `haproxy.router.openshift.io/set-forwarded-headers: never` which then the X-Forwarded-For will be lost as well.

Reply:
@m-yosefpor I'm not sure I understood the problem you faced and whether it's worth being mentioned in the enhancement. Note that you have 4 different policies of how to handle `X-Forwarded-*` headers on OpenShift ingress controller's side: [Using X-Forwarded headers](https://docs.openshift.com/container-platform/4.8/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress).

Reply:
The problem is when somebody uses another port (other than standard 80 and 443), for example 10080 and 10443 for their ingressControllers, the most probable use-case is they have an external loadbalancer in front of routers listening on standard 80 and 443 port, which loadbalances the traffic to 10080 and 10443 ports. For example


11.1.1.1 (external loadbalancer listening on port 443) -> 10.2.1.1:10443 (OCP node IP, with 10443 port specified in `spec.endpointPublishingStrategy.hostNetwork.bindOptions.ports.https` )


A record for ingress: `*.apps.<cluster_name>.<base_domain>.` IN A 11.1.1.1

Then many apps will not work properly, including openshift-image-registry which probably causes e2e tests to fail.

And yes the policy of `X-Forwarded headers` will fix the issue, however we should override the default header policy from `append` to `never` in ingressController, when `spec.endpointPublishingStrategy.hostNetwork.bindOptions` are set to anything other than `http: 80, https: 443`.

```
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpHeaders:
    forwardedHeaderPolicy: Never # <<<<<< this is Append by default which should be overrided by Never, even if user specifies that.
```


The error in case of `append` policy for openshift-image-registry is:

```
Error response from daemon: Get https://image-registry.apps.<cluster_name>.<base_domain>./v2/: Get https://registry.apps.<cluster_name>.<base_domain>.:10443/openshift/token?account=my-sa&client_id=docker&offline_token=true: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers) (Client.Timeout exceeded while awaiting headers)
```

The reason why registry will stop working in this case is that `X-Forwarded-Port` header is set to ``spec.endpointPublishingStrategy.hostNetwork.bindOptions.https` (10443 in this case), will causes a redirect from 443 to 10443 will cause time out (as this port is not even exposed on external LB).

Reply:
Also it is a limitation of listening on non-standard ports, to not to be able to use `append` or `replace` policies, and causes the source IP to be missing, unless we provide other properties in `IngressController`:

e.g. having another separate policy for `X-Forwarded-Port`

```
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpHeaders:
    forwardedHeaderPolicy: Append
    forwardedPortHeaderPolicy: Never # <<< new property
```

Reply:
Hi @m-yosefpor!

Thank you for the investigation of this problem. Sorry for the late reply I was trying to understand why you saw this problem.

1. Image registry indeed uses the `X-Forwarded-*` headers to retrieve the host and port of the original request. This is needed to send the token retrieval link to the original client (`<original-host>:<original-port>/openshift/token?account=my-sa&client_id=docker&offline_token=true`). The code of the image registry [favors the explicit port](https://github.com/openshift/image-registry/blob/master/pkg/origin-common/util/httprequest/httprequest.go#L82) from `X-Forwarded-Port` header over the implicit from `X-Forwarded-Host` and [always gets](https://github.com/openshift/image-registry/blob/master/pkg/origin-common/util/httprequest/httprequest.go#L63) the first value of them.
2. As you know IngressController's default forwardedHeaderPolicy is `Append` which means that `X-Forwarded-*` headers are appended with HAProxy values so that the original values stay the first in the list. In the current HAProxy configuration we append the following x-forwarded headers:
```
X-Forwarded-For
X-Forwarded-Host
X-Forwarded-Port
X-Forwarded-Proto
```

I tried to exercise the behavior of HAProxy using a simple httpserver which responds with all the headers and uses `SchemeHost` implementation from the image registry on OpenShift 4.8.5. Here is the result with simple curl:
```bash
$ curl http://httpserver-default.apps.<public-domain>

Accept=*/*
Forwarded=for=XXX.XXX.XXX.XXX;host=httpserver-default.apps.<public-domain>;proto=http
User-Agent=curl/7.61.1
X-Forwarded-For=XXX.XXX.XXX.XXX
X-Forwarded-Host=httpserver-default.<public-domain>
X-Forwarded-Port=80
X-Forwarded-Proto=http

Now: http httpserver-default.<public-domain>
Before: http httpserver-default.<public-domain>:80
```
Note that `Now` and `Before` outputs are 2 versions of `SchemeHost` function which is [the one](https://github.com/openshift/image-registry/blob/master/pkg/origin-common/util/httprequest/httprequest.go#L59) used by the image registry to get the original host, port and scheme. The current (`Now`) version skips the port if it's 80 for http and 443 for https. 

Here is the output which tried to emulate a LB in between (LB is supposed to insert x-forwarded headers):
```bash
$ curl -H 'X-Forwarded-Host: originaldomain.com' -H 'X-Forwarded-Port: 123' http://httpserver-default.apps.<public-domain>

Accept=*/*
Forwarded=for=XXX.XXX.XXX.XXX;host=httpserver-default.apps.<public-domain>;proto=http
User-Agent=curl/7.61.1
X-Forwarded-For=XXX.XXX.XXX.XXX
X-Forwarded-Host=originaldomain.com,httpserver-default.apps.<public-domain>
X-Forwarded-Port=123,80
X-Forwarded-Proto=http

Now: http originaldomain.com:123
Before: http originaldomain.com:123
```
Note that the original host and port are the first in the list of the values, that's why they appear in `Now` output. This output shows that the image registry would be sending the token request to the original host and port.

I tried a similar exercise on the image registry from the same cluster:
```bash
$ curl -v -k -H 'X-Forwarded-Host: originaldomain.com' -H 'X-Forwarded-Port: 123' https://default-route-openshift-image-registry.apps.<public-domain>/v2/

> GET /v2/ HTTP/1.1
> Host: default-route-openshift-image-registry.apps.<public-domain>
> User-Agent: curl/7.61.1
> Accept: */*
> X-Forwarded-Host: originaldomain.com
> X-Forwarded-Port: 123
> 

< HTTP/1.1 401 Unauthorized
< content-type: application/json; charset=utf-8
< docker-distribution-api-version: registry/2.0
< www-authenticate: Bearer realm="https://originaldomain.com:123/openshift/token"
< x-registry-supports-signatures: 1
< date: Tue, 07 Sep 2021 09:05:43 GMT
< content-length: 87
< 
{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
```
Note the realm in `www-authenticate` header, it's set to the original host and port which I think means that the values appended by HAProxy didn't prevent the image registry from sending the token request to the original client.

So from what I saw, both the ingresscontroller and the imageregistry are implemented with the correct assumptions. If the load balancer which front ends the ingress controller inserts well `X-Forwarded-Host` and `X-Forwarded-Port` headers, they are supposed to be taken into account by the image registry. I assume that the other OpenShift components follow the same logic for the handling of x-forwarded headers as the image registry.

With that said:

- Can you please verify which x-forwarded headers are inserted by the LB you used for the test? Missing `X-Forwarded-Port` header would cause the error you saw (only HAProxy's port would be there in that case).
- You mentioned an e2e test which failed supposedly due to the x-forwarded header problem. I was wondering which e2e you were referring to as the e2e tests from your PRs in `openshift/api` and `openshift/cluster-ingress-operator` were not OKed to be launched yet.

Reply:
Thank you very much @alebedev87 .

> Can you please verify which x-forwarded headers are inserted by the LB you used for the test? Missing X-Forwarded-Port header could cause the error you saw (only HAProxy's port would be there in that case)

Yes I can verify this in our cluster. Actually I used an L4 loadbalancer (hardware Fortigate) for the tests, and there are no `X-Forwarded-..*` headers set by the external LoadBalancer. Setting this header with an L7 LB (or alternatively sending client requests with `X-Forwarded` headers) will resolve the issue.

However the [openshift docs](https://docs.openshift.com/container-platform/4.8/installing/installing_bare_metal/installing-bare-metal.html#installation-load-balancing-user-infra_installing-bare-metal)  mentions that  we should only use L4 LB for app loadbalancing since many other features (such as terminations, etc) might not properly work with L7 ones. Moreover, I suppose other IPI platforms will create some kind of L4 LB setup for `ingressVIP` in `install-config.yaml`, so running ingressController on other host ports will cause this issue for them as well, unless they use proxy-protocol, which is not currently enabled by default ([docs](https://docs.openshift.com/container-platform/4.8/rest_api/operator_apis/ingresscontroller-operator-openshift-io-v1.html) )

> `.spec.endpointPublishingStrategy.hostNetwork` The empty string specifies the default, which is TCP without PROXY protocol. Note that the default is subject to change.


Another point that I should mention, is that image-registry was only an example and many other external apps are using the same mechanism (e.g. KeyCloak) which causes them to misbehaved when deployed on Openshift platforms with this IngressController setup (new non-standard ports).

 
> You mentioned an e2e test which failed supposedly due to the x-forwarded header problem.

I didn't mean the e2e failed already in my PRs in `openshift/api` or `openshift/cluster-ingress-operator` (as even I have not written them yet), but just meant that they will probably fail afterwards.

Reply:
Indeed, L4 LB won't be acting as a reverse proxy and the workloads which rely on `X-Forwarded-*` headers may be misled.
It's rather clear now that the binding on custom ports may have this downside, thanks again for spotting this. 

Let's document this in the enhancement, I'll work with the OpenShift documentation team to have this reflected in there too.

Talking about the workloads, we'd prefer to keep the current approach: append all (port/host/proto) forwarded headers by default. It may be not perfect for all the use cases but unfortunately there is no universal answer for all the scenarios. Generally we rely on the applications to be careful with the dependency on the forwarded headers as they don't always represent a reliable source of truth. Also we provide the configuration options on different levels: ingress and routes which help to tune the policy to the application needs.

Reply:
Is this ok with you, @m-yosefpor to update the enhancement with your findings?

Reply:
@alebedev87 Yes it makes sense. Thank you.

Sorry I was busy this week, I try to work on it and add the explanations in this week.
Please feel free to take over it and add/modify the proposal as you will, especially in case if you have some deadline/time pressure.

Reply:
@m-yosefpor: I created a PR to you fork, it contains mainly the X-Forwarded risk and `bindOptions.ports` field plus some minor changes and rewordings: https://github.com/m-yosefpor/enhancements/pull/1
If you are fine with the changes you can just merge it and they'll appear in this PR.

Reply:
thanks. Merged @alebedev87